### PR TITLE
refactor: help prevent accidental breaking change release

### PR DESCRIPTION
**Pull request by levibostian**
Closes: https://github.com/customerio/issues/issues/6614

This PR's purpose is to introduce a method to help us prevent accidental breaking changes to the SDK without a proper major version bump of our semantic version (`1.0.0` to `2.0.0`). We accomplish this by having .swift files that contain all of the public facing SDK calls. Then, we have an automated tool, [danger](https://danger.systems/js/), that will check to see if any of these .swift files were modified in a pull request. If it finds these files were changed, it will give us a reminder in the pull request to make a proper major version bump to properly introduce a breaking change to the SDK. 

I made some comments below to help explain parts of this PR. 

Complete each step to get your pull request merged in. [Learn more about the workflow this project uses](https://github.com/customerio/customerio-ios/develop/docs/dev-notes/GIT-WORKFLOW.md). 
- [x] [Assign members of your team](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to review the pull request. 
- [x] Wait for pull request status checks to complete. If there are problems, fix them until you see that [all status checks are passing](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fsymfony.com%2Fdoc%2F4.3%2F_images%2Fdocs-pull-request-symfonycloud.png&f=1&nofb=1). 
- [x] Wait until the pull request has been reviewed *and approved* by a teammate
- [ ] After pull request is approved, and you determine it's ready **add the label "Ready to merge"** to the pull request. A bot will *squash and merge* the pull request for you after the label is added.

### CODE REVIEW

File: .github/workflows/pr-review.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  # Danger is a tool that can help during code reviews. Automate common 
+  # code review tasks. https://danger.systems/js/
+  # TL;DR - it runs script `dangerfile.js` on each pull request. 
+  danger:
+    runs-on: ubuntu-latest
+    name: Danger 
+    steps:
+      - uses: actions/checkout@v2
+      - name: Run danger 
+        run: npx danger@11 ci --dangerfile dangerfile.js
```
Comment (levibostian) [inline .github/workflows/pr-review.yml:15]: I added a [new tool](https://danger.systems/js/) to our workflow in this PR. I don't like adding too many different tools to a project to add complexity and require more learning by team members to learn each tool. Luckily, this is a simple tool that simply runs a script on each PR to help us automate parts of the pull request review. 

Right now, the script has only 1 check in it but we can do more in the future.

File: Tests/MessagingPush/APITest.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import SharedTests
+import XCTest
+
+/**
+ Contains an example of every public facing SDK function call. This file helps
+ us prevent introducing breaking changes to the SDK by accident. If a public function
+ of the SDK is modified, this test class will not successfully compile. By not compiling,
+ that is a reminder to either fix the comilation and introduce the breaking change or
+ fix the mistake and not introduce the breaking change in the code base.
+ */
```
Comment (levibostian) [inline Tests/MessagingPush/APITest.swift:13]: This comment on each test file explains what this PR's main purpose is.

File: Tests/Tracking/APITest.swift [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        // Initialize
+        CustomerIO.initialize(siteId: "", apiKey: "")
+        CustomerIO.initialize(siteId: "", apiKey: "", region: .EU)
+
+        // config
+        CustomerIO.config { config in }
+
+        // Identify
+        CustomerIO.shared.identify(identifier: "")
+        // TO FIX: we don't offer a [String: Any] version of identify()
```
Comment (levibostian) [inline Tests/Tracking/APITest.swift:33]: Found a bug while doing this task. 

I'll fix this in [this PR](https://github.com/customerio/issues/issues/6631).

File: Tests/Tracking/APITest.swift [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        // CustomerIO.shared.identify(identifier: "", body: dictionaryData)
+        CustomerIO.shared.identify(identifier: "", body: encodableData)
+
+        // clear identify
+        CustomerIO.shared.clearIdentify()
+
+        // event tracking
+        CustomerIO.shared.track(name: "")
+        // TO FIX: we don't offer a [String: Any] version of track()
```
Comment (levibostian) [inline Tests/Tracking/APITest.swift:42]: Found a bug while doing this task. 

I'll fix this in [this PR](https://github.com/customerio/issues/issues/6631).

File: dangerfile.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+If this pull request *does* introduce a breaking change, make sure the pull request title is in the format: 
+\`\`\` 
+<type>!: description of breaking change 
+// Example:
+refactor!: remove onComplete callback from async functions 
+\`\`\`
+  `
+
+  warn(message)
+})
```
Comment (levibostian) [inline dangerfile.js:20]: This is the script that the tool danger runs on each PR created. This script is written in javascript, yes, but I think the script should read pretty easily. Read the file and see if there are any parts that do not make sense. I want to try and prevent merging in a script that other team members do not understand and feel comfortable contributing to in the future.

File: Tests/MessagingPush/APITest.swift [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+@testable import CioTracking
+import Foundation
+import SharedTests
+import XCTest
+
+/**
+ Contains an example of every public facing SDK function call. This file helps
+ us prevent introducing breaking changes to the SDK by accident. If a public function
+ of the SDK is modified, this test class will not successfully compile. By not compiling,
+ that is a reminder to either fix the comilation and introduce the breaking change or
```
Comment (ami-aman) [inline Tests/MessagingPush/APITest.swift:11]: This comment makes sense to me. On line 11, do you mean "...compilation and introduce..." ?

File: Tests/MessagingPushAPN/APITest.swift [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+@testable import CioTracking
+import Foundation
+import SharedTests
+import XCTest
+
+/**
+ Contains an example of every public facing SDK function call. This file helps
+ us prevent introducing breaking changes to the SDK by accident. If a public function
+ of the SDK is modified, this test class will not successfully compile. By not compiling,
+ that is a reminder to either fix the comilation and introduce the breaking change or
```
Comment (ami-aman) [inline Tests/MessagingPushAPN/APITest.swift:12]: spell check.

File: Tests/Tracking/APITest.swift (hunk 3 above)
Comment (ami-aman) [reply Tests/Tracking/APITest.swift:33]: 👍

**Review by ami-aman — APPROVED** [review]

**Pull request merged**